modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,172 @@
 
 export type FormatChangeCallback = (state: boolean, data: { node: Node; format: string; parents: any }) => void;
 
-export type FormatCallbacks = Record<string, FormatChangeCallback[]>;
+export type RegisteredFormats = Record<string, CallbackGroup>;
 
-export interface FormatData {
-  similar?: boolean;
-  callbacks: FormatChangeCallback[];
+export interface UnbindFormatChanged {
+  unbind: () => void;
 }
 
-export type RegisteredFormats = Record<string, FormatData>;
+interface CallbackWithoutVars {
+  readonly state: Cell<boolean>;
+  readonly similar: boolean;
+  readonly callbacks: FormatChangeCallback[];
+}
 
-export interface UnbindFormatChanged {
-  unbind: () => void;
+interface CallbackWithVars {
+  readonly state: Cell<boolean>;
+  readonly similar: boolean;
+  readonly vars: FormatVars;
+  readonly callback: FormatChangeCallback;
 }
 
-const setup = (registeredFormatListeners: Cell<RegisteredFormats>, editor: Editor) => {
-  const currentFormats = Cell<Record<string, FormatChangeCallback[]>>({ });
+interface CallbackGroup {
+  // Batch all callbacks that don't have custom variables into one "similar=true" and one "similar=false" group
+  readonly withSimilar: CallbackWithoutVars;
+  readonly withoutSimilar: CallbackWithoutVars;
+  // For callbacks with custom variables, handle everything separately
+  readonly withVars: CallbackWithVars[];
+}
 
+const setup = (registeredFormatListeners: Cell<RegisteredFormats>, editor: Editor) => {
   registeredFormatListeners.set({});
 
   editor.on('NodeChange', (e) => {
-    updateAndFireChangeCallbacks(editor, e.element, currentFormats, registeredFormatListeners.get());
+    updateAndFireChangeCallbacks(editor, e.element, registeredFormatListeners.get());
   });
-};
 
-const updateAndFireChangeCallbacks = (editor: Editor, elm: Element, currentFormats: Cell<FormatCallbacks>, formatChangeData: RegisteredFormats) => {
-  const formatsList = Obj.keys(currentFormats.get());
-  const newFormats: FormatCallbacks = { };
-  const matchedFormats: FormatCallbacks = { };
-
-  // Ignore bogus nodes like the <a> tag created by moveStart()
-  const parents = Arr.filter(FormatUtils.getParents(editor.dom, elm), (node) => node.nodeType === 1 && !node.getAttribute('data-mce-bogus'));
-
-  // Check for new formats
-  Obj.each(formatChangeData, (data: FormatData, format: string) => {
-    Tools.each(parents, (node: Node) => {
-      if (editor.formatter.matchNode(node, format, {}, data.similar)) {
-        if (formatsList.indexOf(format) === -1) {
-          // Execute callbacks
-          Arr.each(data.callbacks, (callback: FormatChangeCallback) => {
-            callback(true, { node, format, parents });
-          });
-
-          newFormats[format] = data.callbacks;
-        }
+  editor.on('FormatApply FormatRemove', (e: EditorEvent<FormatEvent>) => {
+    const element = Optional.from(e.node)
+      .map((nodeOrRange) => Obj.get(nodeOrRange as RangeLikeObject, 'startContainer').getOr(nodeOrRange as Node))
```

Comment:
Be aware this may not be working as intended, as `Obj.get` only gets properties owned by the object. Depending on how a native `Range` is implemented, that check may return false and not return the start container. I'd **strongly** suggest changing this to something like this:
```suggestion
      .map((nodeOrRange) => FormatUtils.isNode(node) ? node : node.startContainer)
```
as that is how its handled elsewhere in formatter, so would keep the processing logic consistent.

Reply:
Just confirmed this logic wouldn't have been working with native ranges, which can be passed in `RemoveFormat`:
```
> r = document.createRange()
> Object.hasOwnProperty.call(r, 'startContainer')
false
```